vapi/appliance/simulator/simulator.go

```diff
@@ -0,0 +1,230 @@
+	s.HandleFunc(shell.Path, h.shellStatus)
+}
+
+func (h *Handler) decode(r *http.Request, w http.ResponseWriter, val interface{}) bool {
+	return Decode(r, w, val)
+}
+
+// Decode the request Body into val.
+// Returns true on success, otherwise false and sends the http.StatusInternalServerError response.
+func Decode(request *http.Request, writer http.ResponseWriter, val interface{}) bool {
+	defer request.Body.Close()
+	err := json.NewDecoder(request.Body).Decode(val)
+	if err != nil {
+		log.Printf("%s %s: %s", request.Method, request.RequestURI, err)
+		return false
+	}
+	return true
+}
+
+func (h *Handler) consoleCLIStatus(writer http.ResponseWriter, request *http.Request) {
+	switch request.Method {
+	case http.MethodGet:
+		h.consolecli.RWMutex.RLock()
+		vapi.StatusOK(writer, h.consolecli.access.Enabled)
+		h.consolecli.RWMutex.RUnlock()
+	case http.MethodPut:
+		var input consolecli.Access
+		if h.decode(request, writer, &input) {
+			h.consolecli.RWMutex.Lock()
+			h.consolecli.access.Enabled = input.Enabled
+			h.consolecli.RWMutex.Unlock()
+			writer.WriteHeader(http.StatusNoContent)
+		} else {
+			writer.WriteHeader(http.StatusInternalServerError)
+		}
+	default:
+		http.NotFound(writer, request)
+	}
+}
+
+func (h *Handler) dcuiStatus(writer http.ResponseWriter, request *http.Request) {
+	switch request.Method {
+	case http.MethodGet:
+		h.dcui.RLock()
+		vapi.StatusOK(writer, h.dcui.access.Enabled)
+		h.dcui.RUnlock()
+	case http.MethodPut:
+		var input dcui.Access
+		if h.decode(request, writer, &input) {
+			h.dcui.Lock()
+			h.dcui.access.Enabled = input.Enabled
+			h.dcui.Unlock()
+			writer.WriteHeader(http.StatusNoContent)
+		} else {
+			writer.WriteHeader(http.StatusInternalServerError)
+		}
+	default:
+		http.NotFound(writer, request)
+	}
+}
+
+func (h *Handler) sshStatus(writer http.ResponseWriter, request *http.Request) {
+	switch request.Method {
+	case http.MethodGet:
+		h.ssh.RLock()
+		vapi.StatusOK(writer, h.ssh.access.Enabled)
+		h.ssh.RUnlock()
+	case http.MethodPut:
+		var input ssh.Access
+		if h.decode(request, writer, &input) {
+			h.ssh.Lock()
+			h.ssh.access.Enabled = input.Enabled
+			h.ssh.Unlock()
+			writer.WriteHeader(http.StatusNoContent)
+		} else {
+			writer.WriteHeader(http.StatusInternalServerError)
+		}
+	default:
+		http.NotFound(writer, request)
+	}
+}
+
+func (h *Handler) shellStatus(writer http.ResponseWriter, request *http.Request) {
+	update := func(enabled bool, timeout int) {
+		h.shell.Lock()
+		fmt.Println("Hello")
+		h.shell.access.Enabled = enabled
+		h.shell.access.Timeout = timeout
+		h.shell.Unlock()
+	}
+
+	switch request.Method {
+	case http.MethodGet:
+		h.shell.RLock()
+		vapi.StatusOK(writer, h.shell.access)
+		h.shell.RUnlock()
+	case http.MethodPut:
+		var input shell.Access
+		if h.decode(request, writer, &input) {
+			if h.shell.t != nil {
+				h.shell.t.Stop()
+			}
+			update(input.Enabled, input.Timeout)
+			h.shell.t.Reset(time.Duration(input.Timeout) * time.Second)
```

Comment:
IMHO the timer logic is not correct and can be simplified. Please see the documentation: https://pkg.go.dev/time#NewTimer

Perhaps we can completely remove the timer logic and replace with a goroutine and sleep? Any reason why we need the timer?

Reply:
I introduced the timer logic because of this use-case

- **Step1:** At time 10: enable the shell with a timeout of 20 sec
- **Step2:** At time 10: create a goroutine(1) which resets the shell at time 30
- **Step3:** At time 20: again set the shell with a timeout of 50 Sec
- **Step4:** At time 20: create a goroutine(2) which resets the shell at time 70 sec
- **Step5:** At time 30: goroutine(1) will reset the shell to a disabled state at time 30

This should not be the case, after step 5 the shell will be disabled, this should not be the case. 
with timer logic, we will be doing this

- **Step1:** At time 10: enable the shell with a timeout of 20 sec
- **Step2:** At time 10: create a goroutine(1) which resets the shell at time 30
- **Step3:** At time 20: again set the shell with a timeout of 50 Sec
- **Step4:** At time 20: stop the goroutine(1)
- **Step5:** At time 20: create a goroutine(2) which resets the shell at time 70 sec
- **Step6:** At time 70: goroutine(2) will reset the shell to a disabled state at time 30

Reply:
I think use of a timer vs sleep makes sense for the use-case @embano1 . Sounds like shell access Timeout is a count-down from the[ API doc](https://developer.vmware.com/apis/vsphere-automation/latest/appliance/api/appliance/access/shell/put/) and Timer.Reset can be used to change the Timeout/count-down, unlike Sleep.

Reply:
There were two issues with this code IMHO:

1) it was not correct as per the way of using timer (can be fixed)
2) it was harder to reason about than using a sleep and a bool to indicate whether there is already an existing timer and do a no-op)

I'm fine either way, though correct timer handling in Go adds complexity.

Reply:
Had an offline discussion with @dougm on this and per the discussion, we decided to keep the function simple hence removed timer logic changed the logic to goroutine + sleep as per the suggestion https://github.com/vmware/govmomi/pull/2720#discussion_r801436036 

Current logic is
```
delayReset := func(timeout int) {
		time.AfterFunc(time.Duration(timeout)*time.Second, func() {
			h.shell.access.Enabled = false
			h.shell.access.Timeout = 0
		})
	}

...
if h.decode(request, writer, &input) {
			h.shell.access.Enabled = input.Enabled
			h.shell.access.Timeout = input.Timeout
			go delayReset(input.Timeout)
			writer.WriteHeader(http.StatusNoContent)
		} 
....
```